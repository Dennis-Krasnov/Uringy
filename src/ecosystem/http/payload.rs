//! ...

use ahash::HashMapExt;
use std::cell::OnceCell;
use std::str::FromStr;

#[derive(Debug)]
pub struct Request<'a> {
    method: Method,
    path: &'a str,
    query: &'a str,
    query_map: OnceCell<ahash::HashMap<&'a str, &'a str>>,
    headers: Vec<(&'a str, &'a [u8])>,
    header_map: OnceCell<ahash::HashMap<String, &'a [u8]>>,
    body: &'a [u8],
}

impl<'a> Request<'a> {
    pub(crate) fn new(
        method: Method,
        path: &'a str,
        query: &'a str,
        headers: Vec<(&'a str, &'a [u8])>,
        body: &'a [u8],
    ) -> Self {
        Request {
            method,
            path,
            query,
            query_map: OnceCell::new(),
            headers,
            header_map: OnceCell::new(),
            body,
        }
    }

    /// ...
    #[inline]
    pub fn method(&self) -> Method {
        self.method
    }

    /// ...
    #[inline]
    pub fn path(&self) -> &str {
        &self.path
    }

    /// ...
    #[inline]
    pub fn path_param(&self, _name: &str) -> &str {
        todo!()
    }

    /// ...
    #[inline]
    pub fn raw_query(&self) -> &str {
        &self.query
    }

    /// ... lazy
    #[inline]
    pub fn query_params(&self) -> &ahash::HashMap<&str, &str> {
        self.query_map
            .get_or_init(|| serde_urlencoded::from_str(self.query).unwrap())
    }

    /// ...
    #[inline]
    pub fn query(&self, name: &str) -> Option<&str> {
        self.query_params().get(name).map(|v| *v)
    }

    /// ...
    #[inline]
    pub fn raw_headers(&self) -> &Vec<(&str, &[u8])> {
        &self.headers
    }

    /// ... lazy
    /// ignores duplicates (takes the last)
    #[inline]
    pub fn headers(&self) -> &ahash::HashMap<String, &[u8]> {
        self.header_map.get_or_init(|| {
            let mut map = ahash::HashMap::with_capacity(self.headers.len());
            for (name, value) in &self.headers {
                map.insert(name.to_ascii_lowercase(), *value);
            }
            map
        })
    }

    /// ...
    /// ignores duplicates (takes the last)
    #[inline]
    pub fn header(&self, name: &str) -> Option<&[u8]> {
        self.headers().get(&name.to_ascii_lowercase()).map(|v| *v)
    }

    /// ...
    #[inline]
    pub fn body(&self) -> &[u8] {
        self.body
    }
}

/// ...
#[derive(Debug)]
pub struct Response<'a> {
    pub status: StatusCode,
    pub headers: Vec<(&'a str, &'a [u8])>,
    pub body: &'a [u8],
}

impl Response<'_> {
    /// ...
    #[inline]
    pub fn header(&self, name: &str) -> Option<&[u8]> {
        self.headers
            .iter()
            .find(|(k, _)| k.eq_ignore_ascii_case(name))
            .map(|(_, v)| *v)
    }
}

/// ...
#[derive(Debug, Copy, Clone)]
pub enum Method {
    /// Requests with the GET method:
    /// - Retrieve data at the target resource.
    /// - Shouldn't mutate.
    /// - Shouldn't have a body.
    Get,
    /// Requests with the POST method:
    /// - Submit data to the target resource.
    /// - Aren't idempotent.
    Post,
    /// Requests with the HEAD method:
    /// - Are identical to GET requests, but without the response body.
    Head,
    /// Requests with the PUT method:
    /// - Replace the target resource.
    Put,
    /// Requests with the DELETE method:
    /// - Delete the target resource.
    /// - Shouldn't have a body.
    Delete,
    /// Requests with the CONNECT method:
    /// - Establish a tunnel to the server identified by the target resource.
    /// - Shouldn't have a body.
    Connect,
    /// Requests with OPTIONS method:
    /// - Describe the endpoints the server supports.
    /// - Shouldn't have a body.
    Options,
    /// Requests with the TRACE method:
    /// - Perform a message loop-back test along the path to the target resource.
    /// - Must not have a body.
    Trace,
    /// Requests with the PATCH method:
    /// - Partially update a resource.
    /// - Aren't idempotent.
    Patch,
}

impl FromStr for Method {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "GET" => Ok(Method::Get),
            "POST" => Ok(Method::Post),
            "HEAD" => Ok(Method::Head),
            "PUT" => Ok(Method::Put),
            "DELETE" => Ok(Method::Delete),
            "CONNECT" => Ok(Method::Connect),
            "OPTIONS" => Ok(Method::Options),
            "TRACE" => Ok(Method::Trace),
            "PATCH" => Ok(Method::Patch),
            _ => Err(()),
        }
    }
}

/// ...
#[derive(Debug, Copy, Clone, PartialEq)]
pub enum StatusCode {
    // TODO: rename to Status ???
    Ok,
    Accepted,
    NotModified,
    TemporaryRedirect,
    BadRequest,
    Unauthorized,
    Forbidden,
    NotFound,
    MethodNotAllowed,
}

impl From<StatusCode> for u16 {
    fn from(status: StatusCode) -> Self {
        match status {
            StatusCode::Ok => 200,
            StatusCode::Accepted => 202,
            StatusCode::NotModified => 304,
            StatusCode::TemporaryRedirect => 307,
            StatusCode::BadRequest => 400,
            StatusCode::Unauthorized => 401,
            StatusCode::Forbidden => 403,
            StatusCode::NotFound => 404,
            StatusCode::MethodNotAllowed => 405,
        }
    }
}

/// ...
pub trait AsBody {
    /// ...
    fn contents(&self) -> &[u8];

    /// ...
    fn content_type(&self) -> Option<&str>;
}

impl AsBody for () {
    fn contents(&self) -> &[u8] {
        &[]
    }

    fn content_type(&self) -> Option<&str> {
        None
    }
}

impl AsBody for &str {
    fn contents(&self) -> &[u8] {
        self.as_bytes()
    }

    fn content_type(&self) -> Option<&str> {
        Some("text/plain")
    }
}

impl AsBody for String {
    fn contents(&self) -> &[u8] {
        self.as_bytes()
    }

    fn content_type(&self) -> Option<&str> {
        Some("text/plain")
    }
}

impl AsBody for &[u8] {
    fn contents(&self) -> &[u8] {
        self
    }

    fn content_type(&self) -> Option<&str> {
        Some("application/octet-stream")
    }
}

impl<const N: usize> AsBody for &[u8; N] {
    fn contents(&self) -> &[u8] {
        *self
    }

    fn content_type(&self) -> Option<&str> {
        Some("application/octet-stream")
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn as_body_impls() {
        ().contents();
        "".contents();
        "".as_bytes().contents();
    }

    #[test]
    fn case_insensitive_request_headers() {
        let request = Request::new(
            Method::Get,
            "/",
            "",
            vec![("FOO", b"bar"), ("abc", b"xyz")],
            b"",
        );

        assert_eq!(request.header("foo"), Some("bar".as_bytes()));
        assert_eq!(request.header("ABC"), Some("xyz".as_bytes()));
    }

    // TODO: case_insensitive_response_headers
}
