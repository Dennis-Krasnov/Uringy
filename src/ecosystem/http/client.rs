//! ...

use std::fmt::Debug;
use std::io;
use std::io::{Read, Write};

use http::StatusCode;

use crate::circular_buffer::CircularBuffer;
use crate::ecosystem::http::{Request, Response};

pub fn issue(
    mut connection: impl Read + Write + Debug,
    request: Request,
) -> crate::IoResult<Response> {
    let mut writer = io::BufWriter::new(&mut connection);
    serialize(&mut writer, request)?;
    writer.flush()?;
    let mut connection = writer.into_inner().unwrap();

    deserialize(&mut connection)
}

fn serialize(mut writer: impl Write, request: Request) -> crate::IoResult<()> {
    writer.write_all(request.method().as_str().as_bytes())?;
    writer.write_all(b" ")?;
    writer.write_all(request.uri().to_string().as_bytes())?;
    writer.write_all(b" ")?;
    writer.write_all(format!("{:?}", request.version()).as_bytes())?;
    writer.write_all(b"\r\n")?;

    for (name, value) in request.headers() {
        writer.write_all(name.as_str().as_bytes())?;
        writer.write_all(b": ")?;
        writer.write_all(value.as_bytes())?;
        writer.write_all(b"\r\n")?; // FIXME: still need double \r\n if there's no headers
    }
    writer.write_all(b"\r\n")?;

    let mut body = request.into_body();
    io::copy(&mut body, &mut writer).map_err(crate::Error::from_io_error)?;

    Ok(())
}

fn deserialize(mut reader: impl Read) -> crate::IoResult<Response> {
    let mut buffer = CircularBuffer::new(4096)?;

    loop {
        let bytes_read = reader.read(&mut buffer.uninit())?;
        buffer.commit(bytes_read);

        if bytes_read == 0 {
            panic!("oops"); // TODO: return correct Err
        }

        let mut headers = [httparse::EMPTY_HEADER; 64];
        let mut response = httparse::Response::new(&mut headers);

        match response.parse(&buffer.data()) {
            Ok(httparse::Status::Complete(wire_size)) => {
                let mut builder = Response::builder()
                    .version(http::Version::HTTP_11) // TODO: response.version.unwrap()
                    .status(StatusCode::from_u16(response.code.unwrap()).unwrap());

                let body_size: usize = response
                    .headers
                    .iter()
                    .find(|h| h.name.to_ascii_lowercase() == "content-length")
                    .and_then(|h| std::str::from_utf8(h.value).ok())
                    .and_then(|v| v.parse().ok())
                    .unwrap_or(0);

                for httparse::Header { name, value } in response.headers {
                    builder = builder.header(name.to_string(), value.to_vec());
                }

                if buffer.data().len() < wire_size + body_size {
                    println!("client reading more!");
                    continue;
                }

                let body = buffer.data()[wire_size..(wire_size + body_size)].to_vec();

                buffer.consume(wire_size); // copy from buffer before consuming
                break Ok(builder.raw_body(body).unwrap());
            }
            Ok(httparse::Status::Partial) => continue,
            Err(e) => panic!("oops: {e}"),
        }
    }
}
