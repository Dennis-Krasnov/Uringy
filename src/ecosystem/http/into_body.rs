//! ... request and response.

use std::borrow::Cow;
use std::io::{Cursor, Read};

use mime::Mime;

/// Trait for generating bodies.
pub trait IntoBody {
    // TODO: decide whether this is infallible or not

    /// Create a response.
    /// ... for Content-Length header
    fn into_body(self) -> (Option<usize>, Box<dyn Read>);

    /// ... for Content-Type header
    fn content_type(&self) -> Option<Mime>;
}

impl IntoBody for () {
    fn into_body(self) -> (Option<usize>, Box<dyn Read>) {
        let content_length = Some(0);
        let body = Box::new(Cursor::new(vec![]));
        (content_length, body)
    }

    fn content_type(&self) -> Option<Mime> {
        None
    }
}

impl IntoBody for &'static str {
    fn into_body(self) -> (Option<usize>, Box<dyn Read>) {
        Cow::Borrowed(self).into_body()
    }

    fn content_type(&self) -> Option<Mime> {
        Some(mime::TEXT_PLAIN_UTF_8)
    }
}

impl IntoBody for String {
    fn into_body(self) -> (Option<usize>, Box<dyn Read>) {
        Cow::<'static, str>::Owned(self).into_body()
    }

    fn content_type(&self) -> Option<Mime> {
        Some(mime::TEXT_PLAIN_UTF_8)
    }
}

impl IntoBody for Cow<'static, str> {
    fn into_body(self) -> (Option<usize>, Box<dyn Read>) {
        let content_length = Some(self.len());
        let body = Box::new(Cursor::new(self.as_bytes().to_vec())); // FIXME: don't allocate
        (content_length, body)
    }

    fn content_type(&self) -> Option<Mime> {
        Some(mime::TEXT_PLAIN_UTF_8)
    }
}

impl IntoBody for &'static [u8] {
    fn into_body(self) -> (Option<usize>, Box<dyn Read>) {
        Cow::Borrowed(self).into_body()
    }

    fn content_type(&self) -> Option<Mime> {
        Some(mime::APPLICATION_OCTET_STREAM)
    }
}

impl<const N: usize> IntoBody for &'static [u8; N] {
    fn into_body(self) -> (Option<usize>, Box<dyn Read>) {
        self.as_slice().into_body()
    }

    fn content_type(&self) -> Option<Mime> {
        Some(mime::APPLICATION_OCTET_STREAM)
    }
}

impl<const N: usize> IntoBody for [u8; N] {
    fn into_body(self) -> (Option<usize>, Box<dyn Read>) {
        self.to_vec().into_body()
    }

    fn content_type(&self) -> Option<Mime> {
        Some(mime::APPLICATION_OCTET_STREAM)
    }
}

impl IntoBody for Vec<u8> {
    fn into_body(self) -> (Option<usize>, Box<dyn Read>) {
        Cow::<'static, [u8]>::Owned(self).into_body()
    }

    fn content_type(&self) -> Option<Mime> {
        Some(mime::APPLICATION_OCTET_STREAM)
    }
}

impl IntoBody for Box<[u8]> {
    fn into_body(self) -> (Option<usize>, Box<dyn Read>) {
        Vec::from(self).into_body()
    }

    fn content_type(&self) -> Option<Mime> {
        Some(mime::APPLICATION_OCTET_STREAM)
    }
}

impl IntoBody for Cow<'static, [u8]> {
    fn into_body(self) -> (Option<usize>, Box<dyn Read>) {
        let content_length = Some(self.len());
        let body = Box::new(Cursor::new(self.to_vec())); // FIXME: don't allocate
        (content_length, body)
    }

    fn content_type(&self) -> Option<Mime> {
        Some(mime::APPLICATION_OCTET_STREAM)
    }
}

impl<R: Read + 'static> IntoBody for Box<R> {
    fn into_body(self) -> (Option<usize>, Box<dyn Read>) {
        (None, self)
    }

    fn content_type(&self) -> Option<Mime> {
        todo!()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn compiles() {
        ().into_body();
        "hello".into_body();
        "hello".to_string().into_body();
        b"hello".into_body();
        [b'h', b'i'].into_body();
        b"hello".to_vec().into_body();
        Box::new(Cursor::new("hello")).into_body();
    }
}
