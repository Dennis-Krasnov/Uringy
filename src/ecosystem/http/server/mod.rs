//! ...

use std::io;
use std::io::{Read, Write};

use crate::circular_buffer::CircularBuffer;
use crate::ecosystem::http::server::into_response::IntoResponse;
use crate::ecosystem::http::{Request, Response};

pub mod from_request;
pub mod into_response;
mod macros;
pub mod routing;
// TODO: pub mod middleware

/// ...
pub type Service = Box<dyn Fn(Request) -> Response>;

/// ...
pub trait Handler<ARGS>: Clone {
    /// ...
    fn call(self, request: Request) -> Response;
}

impl<F: FnOnce() -> R + Clone, R: IntoResponse<M>, M> Handler<((), M)> for F {
    fn call(self, _: Request) -> Response {
        self().into_response()
    }
}

macro_rules! impl_handler {
    (
        [$($ty:ident),*], $last:ident
    ) => {
        // include M and variadic arguments in Handler's ARGS to constrain generic type parameters
        #[allow(non_snake_case, unused_mut)]
        impl<F, R, M, M2, $($ty,)* $last> Handler<(M, M2, $($ty,)* $last,)> for F
        where
            F: FnOnce($($ty,)* $last,) -> R + Clone,
            R: $crate::ecosystem::http::server::into_response::IntoResponse<M2>,
            $( $ty: $crate::ecosystem::http::server::from_request::FromRequestParts, )*
            $last: $crate::ecosystem::http::server::from_request::FromRequest<M>,
        {
            fn call(self, request: $crate::ecosystem::http::Request) -> $crate::ecosystem::http::Response {
                let (mut parts, body) = request.into_parts();

                $(
                    let $ty = match $ty::from_request_parts(&mut parts) {
                        Ok(value) => value,
                        Err(rejection) => return rejection.into_response(),
                    };
                )*

                let request = $crate::ecosystem::http::Request::from_parts(parts, body);

                let $last = match $last::from_request(request) {
                    Ok(value) => value,
                    Err(rejection) => return rejection.into_response(),
                };

                self($($ty,)* $last,).into_response()
            }
        }
    };
}

macros::all_the_tuples_and_last!(impl_handler);

fn serialize(mut writer: impl Write, response: Response) -> crate::IoResult<()> {
    writer.write_all(format!("{:?}", response.version()).as_bytes())?;
    writer.write_all(b" ")?;
    writer.write_all(response.status().as_str().as_bytes())?;
    writer.write_all(b" ")?;
    writer.write_all(response.status().canonical_reason().unwrap().as_bytes())?;
    writer.write_all(b"\r\n")?;

    for (name, value) in response.headers() {
        writer.write_all(name.as_str().as_bytes())?;
        writer.write_all(b": ")?;
        writer.write_all(value.as_bytes())?;
        writer.write_all(b"\r\n")?;
    }
    writer.write_all(b"\r\n")?;

    let mut body = response.into_body();
    io::copy(&mut body, &mut writer).map_err(crate::Error::from_io_error)?;

    Ok(())
}

fn deserialize(mut reader: impl Read) -> crate::IoResult<Request> {
    let mut buffer = CircularBuffer::new(4096)?;

    loop {
        let bytes_read = reader.read(&mut buffer.uninit())?;
        buffer.commit(bytes_read);

        if bytes_read == 0 {
            panic!("oops"); // TODO: return correct Err
        }

        let mut headers = [httparse::EMPTY_HEADER; 64];
        let mut request = httparse::Request::new(&mut headers);

        match request.parse(&buffer.data()) {
            Ok(httparse::Status::Complete(wire_size)) => {
                let mut builder = Request::builder()
                    .method(request.method.unwrap())
                    .uri(request.path.unwrap())
                    .version(http::Version::HTTP_11); // TODO: request.version.unwrap()

                let body_size: usize = request
                    .headers
                    .iter()
                    .find(|h| h.name.to_ascii_lowercase() == "content-length")
                    .and_then(|h| std::str::from_utf8(h.value).ok())
                    .and_then(|v| v.parse().ok())
                    .unwrap_or(0);

                for httparse::Header { name, value } in request.headers {
                    builder = builder.header(name.to_string(), value.to_vec());
                }

                if buffer.data().len() < wire_size + body_size {
                    println!("server reading more!");
                    continue;
                }

                let body = buffer.data()[wire_size..(wire_size + body_size)].to_vec();

                buffer.consume(wire_size); // copy from buffer before consuming
                break Ok(builder.raw_body(body).unwrap());
            }
            Ok(httparse::Status::Partial) => continue,
            Err(e) => panic!("oops: {e}"),
        }
    }
}
