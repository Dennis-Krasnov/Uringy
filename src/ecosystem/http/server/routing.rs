//! ...

use std::io::{Read, Write};

use crate::ecosystem::http::server::into_response::IntoResponse;
use crate::ecosystem::http::server::Handler;

/// Handle for composing endpoint handlers.
pub struct Router {
    matcher: matchit::Router<super::Service>,
}

impl Router {
    /// ...
    pub fn new() -> Self {
        Router {
            matcher: matchit::Router::new(),
        }
    }

    /// Add a route to the router.
    pub fn route(mut self, path: &str, method_router: MethodRouter) -> Self {
        // TODO: struct with a field for each method, on handle match request's method
        self.matcher.insert(path, method_router.get).unwrap();
        self
    }

    /// ...
    pub fn handle(&self, request: super::Request) -> super::Response {
        let Ok(endpoint) = self.matcher.at(request.uri().path()) else {
            return http::StatusCode::NOT_FOUND.into_response();
        };

        // endpoint.get("id")

        (endpoint.value)(request)
    }

    /// ...
    pub fn serve(&self, mut connection: impl Read + Write) -> crate::IoResult<()> {
        let request = super::deserialize(&mut connection)?;

        let response = self.handle(request);

        let mut writer = std::io::BufWriter::new(connection);
        super::serialize(&mut writer, response)?;
        writer.flush()?;

        Ok(())
    }
}

/// ...
// TODO: fallback routes
pub struct MethodRouter {
    get: super::Service,
    // post: super::Service,
}

/// Route GET requests to the given handler.
pub fn get<ARGS>(handler: impl Handler<ARGS> + 'static) -> MethodRouter {
    MethodRouter {
        get: Box::new(move |request| handler.clone().call(request)),
    }
}
