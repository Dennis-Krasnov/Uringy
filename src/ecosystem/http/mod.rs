//! ...

pub use http::header;
pub use http::Extensions;
pub use http::Method;
pub use http::StatusCode;
pub use http::Uri;
pub use http::Version;

pub use request::Request;
pub use response::Response;

pub mod client;
pub mod into_body;
pub mod mime;
pub mod request;
pub mod response;
pub mod server;

#[cfg(test)]
mod tests {
    use std::net::Ipv4Addr;

    use crate::ecosystem::http::server::into_response::IntoResponse;
    use crate::ecosystem::http::server::routing::{get, Router};
    use crate::ecosystem::http::{client, Request};
    use crate::net::tcp;
    use crate::runtime::{spawn, start};

    #[test]
    fn end_to_end() {
        start(|| {
            let listener = tcp::Listener::bind((Ipv4Addr::UNSPECIFIED, 0)).unwrap();
            let port = listener.local_addr().unwrap().port();

            let server = spawn(move || {
                let connection = listener.accept().unwrap().0;
                let router = Router::new().route("/", get(root));
                router.serve(connection).unwrap();
            });

            let connection = tcp::Stream::connect((Ipv4Addr::LOCALHOST, port)).unwrap();
            let request = Request::get("/").body("hi").unwrap();
            let response = client::issue(connection, request).unwrap();
            dbg!(&response);
            assert_eq!(response.into_string().unwrap(), "hello");

            server.join().unwrap();
        })
        .unwrap();

        fn root(body: String) -> impl IntoResponse {
            assert_eq!(body, "hi");
            "hello"
        }
    }
}
