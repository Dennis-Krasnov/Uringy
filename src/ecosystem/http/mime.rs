//! body, extractor...

use std::io::{Cursor, Read};
use std::str::FromStr;

use mime::Mime;
use serde::de::DeserializeOwned;
use serde::Serialize;

use crate::ecosystem::http::into_body::IntoBody;
use crate::ecosystem::http::server::from_request::FromRequest;
use crate::ecosystem::http::{header, Request, StatusCode};

/// ...
pub struct Json<T>(pub T);

impl<T: Serialize> IntoBody for Json<T> {
    fn into_body(self) -> (Option<usize>, Box<dyn Read>) {
        let buffer = serde_json::to_vec(&self.0).unwrap(); // FIXME: exception handling
        let content_length = Some(buffer.len());
        let body = Box::new(Cursor::new(buffer));
        (content_length, body)
    }

    fn content_type(&self) -> Option<Mime> {
        Some(mime::APPLICATION_JSON)
    }
}

impl<T: DeserializeOwned> FromRequest for Json<T> {
    type Rejection = (StatusCode, &'static str);

    fn from_request(request: Request) -> Result<Self, Self::Rejection> {
        let is_json = request
            .headers()
            .get(header::CONTENT_TYPE)
            .and_then(|h| h.to_str().ok())
            .and_then(|h| Mime::from_str(h).ok())
            .map_or(false, |mime| {
                mime.type_() == "application"
                    && (mime.subtype() == "json"
                        || mime.suffix().map_or(false, |name| name == "json"))
            });

        if !is_json {
            return Err((
                StatusCode::UNSUPPORTED_MEDIA_TYPE,
                "Expected request with `Content-Type: application/json`",
            ));
        }

        let body = request.into_body();
        let value = serde_json::from_reader(body).map_err(|e| match e.classify() {
            serde_json::error::Category::Io => (StatusCode::INTERNAL_SERVER_ERROR, ""),
            serde_json::error::Category::Syntax | serde_json::error::Category::Eof => (
                StatusCode::BAD_REQUEST,
                "Failed to parse the request body as JSON`",
            ),
            serde_json::error::Category::Data => (
                StatusCode::BAD_REQUEST,
                "Failed to deserialize the JSON body into the target type",
            ),
        })?;
        Ok(Json(value))
    }
}

/// ...
pub struct Html<T>(pub T);

impl<T: IntoBody> IntoBody for Html<T> {
    fn into_body(self) -> (Option<usize>, Box<dyn Read>) {
        self.0.into_body()
    }

    fn content_type(&self) -> Option<Mime> {
        Some(mime::TEXT_HTML_UTF_8)
    }
}

/// ...
pub struct JavaScript<T>(pub T);

impl<T: IntoBody> IntoBody for JavaScript<T> {
    fn into_body(self) -> (Option<usize>, Box<dyn Read>) {
        self.0.into_body()
    }

    fn content_type(&self) -> Option<Mime> {
        Some(mime::APPLICATION_JAVASCRIPT_UTF_8)
    }
}

/// ...
pub struct Css<T>(pub T);

impl<T: IntoBody> IntoBody for Css<T> {
    fn into_body(self) -> (Option<usize>, Box<dyn Read>) {
        self.0.into_body()
    }

    fn content_type(&self) -> Option<Mime> {
        Some(mime::TEXT_CSS_UTF_8)
    }
}

#[cfg(test)]
mod tests {
    use crate::ecosystem::http::server::routing::{get, Router};
    use crate::ecosystem::http::{Request, Response, StatusCode};

    use super::*;

    mod json {
        use super::*;

        #[test]
        fn takes_valid() {
            fn root(Json(number): Json<i32>) {
                assert_eq!(number, 42);
            }
            let router = Router::new().route("/", get(root));

            let request = Request::get("/").body(Json(42)).unwrap();
            let response = router.handle(request);

            assert_eq!(response.status(), &StatusCode::OK);
        }

        #[test]
        fn cant_take_when_missing_content_type() {
            fn root(_: Json<i32>) {
                unreachable!();
            }
            let router = Router::new().route("/", get(root));

            let request = Request::get("/").body(()).unwrap();
            let response = router.handle(request);

            assert_eq!(response.status(), &StatusCode::UNSUPPORTED_MEDIA_TYPE);
        }

        #[test]
        fn cant_take_when_invalid_content_type() {
            fn root(_: Json<i32>) {
                unreachable!();
            }
            let router = Router::new().route("/", get(root));

            let request = Request::get("/").body("42").unwrap();
            let response = router.handle(request);

            assert_eq!(response.status(), &StatusCode::UNSUPPORTED_MEDIA_TYPE);
        }

        #[test]
        fn cant_take_when_invalid_value() {
            fn root(_: Json<i32>) {
                unreachable!();
            }
            let router = Router::new().route("/", get(root));

            let request = Request::get("/").body(Json("42")).unwrap();
            let response = router.handle(request);

            assert_eq!(response.status(), &StatusCode::BAD_REQUEST);
        }

        #[test]
        fn is_into_body() {
            let request = Request::builder().body(Json(42)).unwrap();
            assert!(request.headers()[header::CONTENT_TYPE]
                .to_str()
                .unwrap()
                .contains("json"));

            let response = Response::builder().body(Json(42)).unwrap();
            assert!(response.headers()[header::CONTENT_TYPE]
                .to_str()
                .unwrap()
                .contains("json"));
        }
    }

    #[test]
    fn html_is_into_body() {
        let request = Request::builder().body(Html("hello")).unwrap();
        assert!(request.headers()[header::CONTENT_TYPE]
            .to_str()
            .unwrap()
            .contains("html"));

        let response = Response::builder().body(Html("hello")).unwrap();
        assert!(response.headers()[header::CONTENT_TYPE]
            .to_str()
            .unwrap()
            .contains("html"));
    }

    #[test]
    fn javascript_is_into_body() {
        let request = Request::builder().body(JavaScript("hello")).unwrap();
        assert!(request.headers()[header::CONTENT_TYPE]
            .to_str()
            .unwrap()
            .contains("javascript"));

        let response = Response::builder().body(JavaScript("hello")).unwrap();
        assert!(response.headers()[header::CONTENT_TYPE]
            .to_str()
            .unwrap()
            .contains("javascript"));
    }

    #[test]
    fn css_is_into_body() {
        let request = Request::builder().body(Css("hello")).unwrap();
        assert!(request.headers()[header::CONTENT_TYPE]
            .to_str()
            .unwrap()
            .contains("css"));

        let response = Response::builder().body(Css("hello")).unwrap();
        assert!(response.headers()[header::CONTENT_TYPE]
            .to_str()
            .unwrap()
            .contains("css"));
    }
}
