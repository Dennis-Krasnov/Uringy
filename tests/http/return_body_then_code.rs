use http::StatusCode;
use uringy::ecosystem::http::server::routing::{get, Router};

fn main() {
    Router::new().route("/", get(|| ("hello", StatusCode::OK)));
}
