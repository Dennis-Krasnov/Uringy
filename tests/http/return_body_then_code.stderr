error[E0277]: the trait bound `[closure@$DIR/tests/http/return_body_then_code.rs:5:34: 5:36]: Handler<_>` is not satisfied
 --> tests/http/return_body_then_code.rs:5:34
  |
5 |     Router::new().route("/", get(|| ("hello", StatusCode::OK)));
  |                              --- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Handler<_>` is not implemented for closure `[closure@$DIR/tests/http/return_body_then_code.rs:5:34: 5:36]`
  |                              |
  |                              required by a bound introduced by this call
  |
note: required by a bound in `uringy::ecosystem::http::server::routing::get`
 --> src/ecosystem/http/server/routing.rs
  |
  | pub fn get<ARGS>(handler: impl Handler<ARGS> + 'static) -> MethodRouter {
  |                                ^^^^^^^^^^^^^ required by this bound in `get`
